### DISCUSSION TITLE
JSON updates with $ - Expected an identifier by Dart

### DISCUSSION THREAD

**Discussion by tiofabby** [ra1u/redis-dart#63] [Q&A]
Hi,
I am trying to update a JSON entry with the following command:
`await Command.send_object((["JSON.SET", "myKey", "$.myfield", newValue]));
`
Issue is that "$.myfield" is not accepted by dart, it says **"Expected an Identifier"** and so it does not compile.
I tried to put $ alone like: "$", "json format.." but Dart still complains.. it does not like having a $.

Would someone know how to do a JSON UPDATE then using command.send_object ?

Thank you
Fabrice

**Comment by ra1u** [comment]
I think that it is string in dart that has an issue. Can you try with raw strings and report if that heps.

    await Command.send_object((["JSON.SET", "myKey", r"$.myfield", newValue]));
    
 [String reference](https://dart.dev/guides/language/language-tour#strings)

**Comment by tiofabby** [comment]
Hi @ra1u , Thank you for your quick feedback. 
I tried what you suggested, it compiles correctly now but it gives an exception when executed:
I/flutter ( 8131): The following exception was thrown Error in ...:
I/flutter ( 8131): RedisError(expected value at line 1 column 1)

**Reply by ra1u** [reply]
I suspect that _newValue_ is not json in form expected by  redis. Can you provide example that can be reproduced and would be clear what _newValue_ is?

**Reply by tiofabby** [reply]
Sure, here is full line with your "r" suggestion (my keys are "player:uid" and I am trying to update the avatar string field (image name) of the user):
```
await newCommand.send_object(([
          "JSON.SET",
          "player:${user.userId}",
          r"$.avatar",
          "${user.avatar}"
        ]));

```
It works just fine if I do 
```
await newCommand.send_object(([
          "JSON.SET",
          "player:${user.userId}",
          ".",
          '{ "avatar": "${user.avatar}"}'
        ]));
``` 
but the issue is that it does not do an update it replaces the whole JSON entry with this new one. In my case, I am looking for a way of just updating one field and keep other ones.
Thanks

**Reply by ra1u** [reply]
I see. It think that you not using redis set as it should be. Take a look into example as documeted by redis in  [JSON.SET](https://oss.redis.com/redisjson/commands/#jsonset) .  Can you check that what you are trying to do complies with command that can be executed over terminal as shown in example.

**Reply by tiofabby** [reply]
Trying to make it work manually with RedisInsight it fails too and I ended finding this [link](https://github.com/RedisJSON/RedisJSON/issues/34) which kind of say that we cannot update just one field in one JSON entry... I feel this is a pity but I would have to deal with such bad behavior I guess...

**Reply by craiglabenz** [reply]
@tiofabby, try this:

```dart
await newCommand.send_object(([
          "JSON.SET",
          "player:${user.userId}",
          r"$.avatar",
          '"${user.avatar}"'
        ]));
```
The difference changing `"${user.avatar}"` into `'"${user.avatar}"'`. You need quotes INSIDE the string you pass to Redis.
